### PR TITLE
[BUG] Fix redirect-from Sphinx extension

### LINKED ISSUE #21876: [Doc]: redirect-from directive appears broken?
### Documentation Link

https://matplotlib.org/devdocs/

### Problem

We have a number of `.. redirect-from::` directives that are supposed to create a redirect html in the old location.  That appears to have broken. 

ie. `doc/users/faq/troubleshooting_faq.rst` has a `.. redirect-from:: /faq/troubleshooting_faq` which is _supposed_ to make a file https://matplotlib.org/devdocs/faq/troubleshooting_faq.html that redirects to https://matplotlib.org/devdocs/users/faq/troubleshooting_faq.html  However that file is not being created and so far as I can see no error messages are being emitted.  

This is loaded from `doc/sphinxext/redirect_from.py`

### Suggested improvement

This should be fixed

### Matplotlib Version

main

### LINKED ISSUE #27911: redirect not working for updated contribute page
There should be a re-direct there: https://github.com/matplotlib/matplotlib/blob/478dc5aa62d90f9abeab62b94aefc1a1bfde3413/doc/devel/contribute.rst?plain=1#L1-L3

We should merge this to avoid the redirect in our links, but we should also sort out why the redirect is not working.

_Originally posted by @tacaswell in https://github.com/matplotlib/matplotlib/issues/27910#issuecomment-1992091204_

### PR DESCRIPTION
**Pull request by AnsonTran**
## PR summary

Closes #27911, closes #21876. This PR fixes the issue of redirect-from directives not
creating the relative refresh pages in the build.

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [N/A] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [N/A] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [N/A] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [N/A] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Review by timhoffm — COMMENTED** [review]
The `clear_doc` change to operating on a single document is correct. Clearing everything was a mistake.

Why do we have to delete the consistency checks?

**Comment by AnsonTran** [comment]
If the docs build successfully the first time with no changes, Sphinx will load in a cached build environment with ```domain.requests``` already populated. Hence, that check will cause the build to fail

**Comment by tacaswell** [comment]
> Sphinx will load in a cached build environment with domain.requests already populated. Hence, that check will cause the build to fail

But that suggests the checks are doing their job and we still have a bug?

**Comment by AnsonTran** [comment]
Added the sanity checks back in. Also added a function to clear the domain redirects, so build doesnt fail when rebuilding without doing a ```make clean```.

File: doc/sphinxext/redirect_from.py [hunk 1]
```diff
@@ -73,28 +73,29 @@ def redirects(self):
...
     def merge_domaindata(self, docnames, otherdata):
-        for src, dst in otherdata['redirects'].items():
-            if src not in self.redirects:
-                self.redirects[src] = dst
-            elif self.redirects[src] != dst:
+        for docname in docnames:
+            otherdoc = otherdata['redirects'][docname]
+            if self.redirects[docname]:
+                src, dst = self.redirects[docname]
+                _, otherdst = otherdoc
```
Comment (timhoffm) [inline doc/sphinxext/redirect_from.py:84]: Can you please give the motivation for this change? I find this quite confusing, but I belive this is wrong. AFAICS from l.194 (`domain.redirects[redirected_reldoc] = current_doc`) `self.redirects` is a mapping "old (non-existing) page" --> "new (existing) page". OTOH the parameter `docnames` contains the existing pages. So, I belive it is incorrect to use it as a key in `self.redirects`.

Comment (AnsonTran) [reply doc/sphinxext/redirect_from.py:84]: Hi Tim, the original motivation for the change was to match the implementation here: 
https://github.com/sphinx-doc/sphinx/blob/5523c9bbe4d1415777669330411d4f00ad802f18/sphinx/ext/todo.py#L90-L92

I did make a mistake when adding the sanity check for this function, so I will undo the changes and leave it as is.

**Review by timhoffm — APPROVED** [review]

File: doc/sphinxext/redirect_from.py [hunk 2]
```diff
@@ -119,3 +118,10 @@ def _generate_redirects(app, exception):
             logger.info('making refresh html file: %s redirect to %s', k, v)
             p.parent.mkdir(parents=True, exist_ok=True)
             p.write_text(html, encoding='utf-8')
+
+
+def _clear_redirects(app):
+    domain = app.env.get_domain('redirect_from')
+    if (domain.redirects):
+        domain.redirects.clear()
+        logger.info('clearing cached redirects')
```
Comment (QuLogic) [inline doc/sphinxext/redirect_from.py:127]: Unnecessary parentheses. Also, the log message is written to go first.
```suggestion
    if domain.redirects:
        logger.info('clearing cached redirects')
        domain.redirects.clear()
```

Comment (AnsonTran) [reply doc/sphinxext/redirect_from.py]: Implemented, thank you!

**Review by QuLogic — APPROVED** [review]

**Pull request merged**